### PR TITLE
feat: allow manipulation of CoinJoin salt using RPC

### PR DESCRIPTION
**Pull request by kwvg**
## Motivation

CoinJoin utilizes a salt to randomize the number of rounds used in a mixing session (introduced in [dash#3661](https://github.com/dashpay/dash/pull/3661)). This was done to frustrate attempts at deobfuscating CoinJoin mixing transactions but also has the effect of deciding the mixing threshold at which a denomination is considered "sufficiently mixed", which in turn is tied this salt, that is in turn, tied to the wallet.

With wallets that utilized keypool generation, this was perfectly acceptable as each key was unrelated to the other and any meaningful attempts to backup/restore a wallet would entail backing up the entire wallet database, which includes the salt. Meaning, on restore, the wallet would show CoinJoin balances as reported earlier.

With the default activation of HD wallets in legacy wallets (and the introduction of descriptor wallets that construct HD wallets by default), addresses are deterministically generated and backups no longer _require_ a backup of the wallet database wholesale.

Users who export their mnemonic and import them into a new wallet will find themselves with a different reported CoinJoin balance (see below).

https://github.com/dashpay/dash/assets/63189531/dccf1b17-55af-423d-8c36-adea6163e060

## Demo

**Based on [`c00a3665`](https://github.com/kwvg/dash/commit/c00a366578354810908e4034c5a73ab3782f2cc6)**

https://github.com/dashpay/dash/assets/63189531/c60c10e3-e414-46af-a64e-60605a4e6d07

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2757,21 +2757,36 @@ const CTxOut& CWallet::FindNonChangeParentOutput(const CTransaction& tx, int out
...
     }
 
-    while (nCoinJoinSalt.IsNull()) {
-        // We never generated/saved it
-        nCoinJoinSalt = GetRandHash();
-        batch.WriteCoinJoinSalt(nCoinJoinSalt);
+    return nCoinJoinSalt.IsNull();
+}
+
+bool CWallet::SetCoinJoinSalt(const uint256 cj_salt)
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/wallet/wallet.cpp]: Resolved in latest push!

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -2757,21 +2757,36 @@ const CTxOut& CWallet::FindNonChangeParentOutput(const CTransaction& tx, int out
...
+{
+    WalletBatch batch(GetDatabase());
+    // Only store new salt in CWallet if database write is successful
+    if (batch.WriteCoinJoinSalt(nCoinJoinSalt)) {
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/wallet/wallet.cpp]: Resolved in latest push!

File: src/rpc/coinjoin.cpp [hunk 3]
```diff
@@ -177,9 +405,13 @@ static RPCHelpMan getcoinjoininfo()
...
     manager->GetJsonInfo(obj);
 
-    obj.pushKV("keys_left",     wallet->nKeysLeftSinceAutoBackup);
-    obj.pushKV("warnings",      wallet->nKeysLeftSinceAutoBackup < COINJOIN_KEYS_THRESHOLD_WARNING
-                                        ? "WARNING: keypool is almost depleted!" : "");
+    std::string warning_msg{""};
+    if (wallet->IsLegacy()) {
+        obj.pushKV("keys_left", wallet->nKeysLeftSinceAutoBackup);
+        warning_msg = (wallet->nKeysLeftSinceAutoBackup < COINJOIN_KEYS_THRESHOLD_WARNING) 
+                        ? "WARNING: keypool is almost depleted!" : "";
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/rpc/coinjoin.cpp]: Resolved in latest push!

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -2757,21 +2757,36 @@ const CTxOut& CWallet::FindNonChangeParentOutput(const CTransaction& tx, int out
...
-void CWallet::InitCoinJoinSalt()
+const uint256& CWallet::GetCoinJoinSalt()
+{
+    if (nCoinJoinSalt.IsNull()) {
+        ReadCoinJoinSalt();
+    }
+    return nCoinJoinSalt;
+}
+
+bool CWallet::ReadCoinJoinSalt()
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/wallet/wallet.cpp]: Didn't want to use `InitCoinJoinSalt` as existing behaviour dictates that if there is no salt, then one will be generated and committed but as you mentioned, using `ReadCoinJoinSalt` is misleading as there _may_ be write operations, named it `InitCJSaltFromDb` to make it clearer. Resolved in latest push!

**Comment by UdjinM6** [comment]
Let's split it in 3 separate PRs:
- cj salt manipulation
- make `coinjoin` a composite command
- fixing `getcoinjoininfo` for descriptor wallets

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
I think I have a solution how to avoid (very slow) rescans, pls check 78625f80e3

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -1448,6 +1448,15 @@ int CWallet::GetCappedOutpointCoinJoinRounds(const COutPoint& outpoint) const
     return realCoinJoinRounds > CCoinJoinClientOptions::GetRounds() ? CCoinJoinClientOptions::GetRounds() : realCoinJoinRounds;
 }
 
+void CWallet::ClearCoinJoinRoundsCache()
+{
+    LOCK(cs_wallet);
+    mapOutpointRoundsCache.clear();
+    MarkDirty();
+    // Notify UI
+    NotifyTransactionChanged(uint256::ONE, CT_UPDATED);
```
Comment (knst) [inline src/wallet/wallet.cpp:1457]: should NotifyTransactionChanges be out of scope of LOCK of cs_wallet?
@UdjinM6

Comment (UdjinM6) [reply src/wallet/wallet.cpp:1457]: `cs_wallet ` is held in all other use cases, should be fine here too imo.

**Review by UdjinM6 — APPROVED** [review]
ACK 019a3c070d50f97260cc1e2462c36274003e1b42

**Review by UdjinM6 — APPROVED** [review]
rebase looks clean

utACK de50cad86c8f30c82aa73b3d8e152f60bc549ca2

**Review by PastaPastaPasta — COMMENTED** [review]
generally LGTM; want @knst input on the composite command stuff and @thephez input on the user facing messages

File: src/rpc/coinjoin.cpp [hunk 6]
```diff
@@ -181,6 +181,188 @@ static RPCHelpMan coinjoin_stop()
...
+{
+    throw JSONRPCError(RPC_INVALID_PARAMETER, "Must be a valid command");
+},
+    };
+}
+
+static RPCHelpMan coinjoinsalt_generate()
+{
+    return RPCHelpMan{"coinjoinsalt generate",
+        "\nGenerate new CoinJoin salt and commit to wallet database\n"
```
Comment (thephez) [inline src/rpc/coinjoin.cpp:210]:
```suggestion
        "\nGenerate new CoinJoin salt and store it in the wallet database\n"
```

File: src/rpc/coinjoin.cpp [hunk 7]
```diff
@@ -181,6 +181,188 @@ static RPCHelpMan coinjoin_stop()
...
+        },
+        [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+{
+    std::shared_ptr<CWallet> const wallet = GetWalletForJSONRPCRequest(request);
+    if (!wallet) return NullUniValue;
+
+    const auto salt{ParseHashV(request.params[0], "salt")};
+    const bool enable_overwrite{!request.params[1].isNull() ? request.params[1].get_bool() : /* default */ false};
+    if (salt == uint256::ZERO) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Illegal CoinJoin salt value");
```
Comment (thephez) [inline src/rpc/coinjoin.cpp:328]: Nothing to see here officer :oncoming_police_car: 
Invalid seems like a better option...
```suggestion
        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid CoinJoin salt value");
```

**Review by thephez — CHANGES_REQUESTED** [review]
A couple suggestions @PastaPastaPasta @kwvg

File: src/rpc/coinjoin.cpp (hunk 6 above)
Comment (kwvg) [reply src/rpc/coinjoin.cpp]: Resolved in latest push!

File: src/rpc/coinjoin.cpp (hunk 7 above)
Comment (kwvg) [reply src/rpc/coinjoin.cpp]: Resolved in latest push!

**Review by thephez — APPROVED** [review]
utACK

**Review by UdjinM6 — DISMISSED** [review]
utACK a467f7ae46a62f5abfab331b026d8a1740a8954e

**Review by knst — COMMENTED** [review]
@kwvg could you try to rebase your PR on top of https://github.com/dashpay/dash/pull/6110 ?

that PR removes linter `test/lint/check-rpc-mappings.py` and many other changes/improvements

File: src/rpc/coinjoin.cpp [hunk 8]
```diff
@@ -181,6 +181,188 @@ static RPCHelpMan coinjoin_stop()
...
+            HelpExampleCli("coinjoinsalt set", "f4184fc596403b9d638783cf57adfe4c75c605f6356fbc91338530e9831e9e16")
+          + HelpExampleRpc("coinjoinsalt set", "f4184fc596403b9d638783cf57adfe4c75c605f6356fbc91338530e9831e9e16")
+        },
+        [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+{
+    std::shared_ptr<CWallet> const wallet = GetWalletForJSONRPCRequest(request);
+    if (!wallet) return NullUniValue;
+
+    const auto salt{ParseHashV(request.params[0], "salt")};
+    const bool enable_overwrite{!request.params[1].isNull() ? request.params[1].get_bool() : /* default */ false};
```
Comment (knst) [inline src/rpc/coinjoin.cpp:326]: could you exclude composite command refactorings or split it to new PR?

Composite commands's related changes are conflicting with #6110 

Instead, you can use (at least temporary) other approach to get a params for composite commands:
```diff
-const bool enable_overwrite{!request.params[0].isNull() ? request.params[0].get_bool() : /* default */ false};
+const bool enable_overwrite = false;
+if (!request.params[0].isNull()) {
+    enable_overwrite = ParseBoolV(request.params[0], "overwrite");
+}
```

**Comment by kwvg** [comment]
Marking as draft until #6110 is merged

File: src/rpc/coinjoin.cpp (hunk 8 above)
Comment (kwvg) [reply src/rpc/coinjoin.cpp]: Resolved in latest push

**Review by knst — APPROVED** [review]
utACK c52c9f86df0873f880a6dbab24300c3b7bbf1cc4


could you add a functional test for `coinjoinsalt`? You may create new PR, if you feel it is out-of-scope of this PR, but it have to be done for sack of enabling rpc-coverage linter which is not running due to https://github.com/dashpay/dash-issues/issues/63

File: src/rpc/coinjoin.cpp [hunk 9]
```diff
@@ -181,6 +181,196 @@ static RPCHelpMan coinjoin_stop()
...
+    const bool has_balance{(wallet_balance.m_anonymized
+                          + wallet_balance.m_denominated_trusted
+                          + wallet_balance.m_denominated_untrusted_pending) > 0};
+    if (has_balance && !enable_overwrite) {
+        throw JSONRPCError(RPC_WALLET_ERROR,
+                           strprintf("Wallet \"%s\" has CoinJoin balance, cannot continue!", str_wallet));
+    }
+
+    if (!wallet->SetCoinJoinSalt(salt)) {
+        throw JSONRPCError(RPC_INVALID_REQUEST,
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/rpc/coinjoin.cpp]: Resolved in latest push

**Review by knst — APPROVED** [review]
re-utACK https://github.com/dashpay/dash/commit/c52c9f86df0873f880a6dbab24300c3b7bbf1cc4

could you add a functional test for coinjoinsalt? You may create new PR, if you feel it is out-of-scope of this PR, but it have to be done for sack of enabling rpc-coverage linter which is not running due to https://github.com/dashpay/dash-issues/issues/63

**Comment by kwvg** [comment]
> could you add a functional test for coinjoinsalt?

Will do in a follow-up

**Review by PastaPastaPasta — APPROVED** [review]
utACK 8e5e293d4ca36a9b88731648c7581d430173b05e

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: src/rpc/coinjoin.cpp [hunk 10]
```diff
@@ -181,6 +181,196 @@ static RPCHelpMan coinjoin_stop()
...
+    if (!wallet) return NullUniValue;
+
+    const auto salt{ParseHashV(request.params[0], "salt")};
+    if (salt == uint256::ZERO) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid CoinJoin salt value");
+    }
+
+    bool enable_overwrite{false}; // Default value
+    if (!request.params[0].isNull()) {
+        enable_overwrite = ParseBoolV(request.params[0], "overwrite");
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/rpc/coinjoin.cpp]: Good catch, resolved in latest push!

File: src/rpc/coinjoin.cpp [hunk 11]
```diff
@@ -181,6 +181,196 @@ static RPCHelpMan coinjoin_stop()
 },
     };
 }
+
+// From wallet/rpcdump.cpp
+extern void RescanWallet(CWallet& wallet, const WalletRescanReserver& reserver, int64_t time_begin = 0, bool update = true);
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/rpc/coinjoin.cpp]: Resolved in latest push!

File: src/wallet/rpcdump.cpp [hunk 12]
```diff
@@ -59,7 +59,7 @@ static std::string DecodeDumpString(const std::string &str) {
 
 static const int64_t TIMESTAMP_MIN = 0;
 
-static void RescanWallet(CWallet& wallet, const WalletRescanReserver& reserver, int64_t time_begin = TIMESTAMP_MIN, bool update = true)
+void RescanWallet(CWallet& wallet, const WalletRescanReserver& reserver, int64_t time_begin = TIMESTAMP_MIN, bool update = true)
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/wallet/rpcdump.cpp]: Resolved in latest push!

**Review by knst — APPROVED** [review]
re-utACK c52c9f86df0873f880a6dbab24300c3b7bbf1cc4

**Review by UdjinM6 — APPROVED** [review]
ACK 5943c93bdd36a7c4ee96e94d60772814111f7cd8

**Comment by knst** [comment]
@kwvg I just realized, there's no functional tests for it. Our linter complains about it:

```
$ test/functional/test_runner.py -j20  --previous-releases --coverage --extended\
.....
Uncovered RPC commands:
  - cleardiscouraged
  - coinjoinsalt <-------
  - debug
  - getblockheaders
  - getmerkleblocks
  - getpoolinfo
  - voteraw
```

See also https://github.com/dashpay/dash-issues/issues/63

**Pull request merged**